New zimdump cmdline arguments.

MiguelRocha opened the pull request:
This PR refactors the argument parsing mechanism of zimdump. It applies a subcommand structure with the current subcommands:

- info
- dumpall
- dump
- list
- verifychecksum

data-man commented:
IMO [CLI11](https://github.com/CLIUtils/CLI11) is the best command line parser I ever tried.

kelson42 commented:
@mgautierfr @MiguelRocha I think the `verifychecksum` action should/could be removed in favour of the one in `zimcheck`. What do you think?

kelson42 commented on src/zimdump.cpp:
```diff
@@ -290,114 +298,100 @@ void ZimDumper::verifyChecksum()
+      zimdump <zim_file> list [--asTable|--metadata] [--extra]
+      zimdump <zim_file> dump [--url=<urlvalue>|--title=<titlevalue>|--offset=<offsetvalue>] [--namespace=<ns>] [--asPage|--asMeta] [--extra]
+      zimdump <zim_file> verifychecksum
+
```
To me there is an empty line too much here.

kelson42 commented:
I'm a bit skeptical here because:
* `zimdump wikipedia_en_ray_charles_maxi_2020-03.zim list --namespace=X` seems to not work (and I don't understand why?)
* It seems there is no option at all to find an article
* Most of the options are not clear to me (unexplained).
* `--asTable` does not work as it should (this is a CSV, not a table)

Probably (part of) the problem comes from an incomplete/weak requirements :(

mgautierfr commented on meson.build:
```diff
@@ -16,5 +16,6 @@ libzim_dep = dependency('libzim', version : '>=6.1.0', static:static_linkage)
 
 compiler = meson.get_compiler('cpp')
 rt_dep = compiler.find_library('rt', required:false)
+docopt_dep = compiler.find_library('docopt', required:false)
```
docopt is required.

mgautierfr commented on src/zimdump.cpp:
```diff
@@ -290,114 +298,99 @@ void ZimDumper::verifyChecksum()
+      zimdump <zim_file> dumpall --output=<outputdir>
+      zimdump <zim_file> list [--asTable|--metadata] [--extra]
+      zimdump <zim_file> dump [--url=<urlvalue>|--title=<titlevalue>|--offset=<offsetvalue>] [--namespace=<ns>] [--asPage|--asMeta] [--extra]
+      zimdump <zim_file> verifychecksum
```
We shouldn't have a verifychecksum sub-command.
zimcheck is better for that.

mgautierfr commented:
> zimdump wikipedia_en_ray_charles_maxi_2020-03.zim list --namespace=X seems to not work (and I don't understand why?)

There is not option `--namespace` for sub-command `list` so you cannot use it.

> --asTable does not work as it should (this is a CSV, not a table)

It is how "previous" zimdump was working.

MiguelRocha commented:
@mgautierfr do you agree with the implementation? Can you please review the code?

kelson42 commented:
@MiguelRocha The whole CI is failing, the discussions are not resumed. Might you please clean the situation?

mgautierfr commented on src/zimdump.cpp:
```diff
@@ -393,124 +402,128 @@ void ZimDumper::verifyChecksum()
 
-int main(int argc, char* argv[])
+static const char USAGE[] =
+R"(zimdump.
+
+    Usage:
+      zimdump <zim_file> info [--ns=N]
+      zimdump <zim_file> list [--details] [--idx=INDEX|(--url=URL [--ns=N])]
+      zimdump <zim_file> dump --dir=DIR [--ns=N] [--redirect]
+      zimdump <zim_file> dump (--idx=INDEX |(--url=URL [--ns=N]))
+
+    Options:
+      --idx=INDEX       The index of the article to list/dump
+      --ns=N            Namespace selection [default: A]
+      -h --help         Show this screen.
+      --version         Show version.
+)";
```
Please use the usage text I've give in https://github.com/openzim/zim-tools/issues/70#issuecomment-636773620

mgautierfr commented on src/zimdump.cpp:
```diff
@@ -393,124 +402,128 @@ void ZimDumper::verifyChecksum()
+#ifdef _WIN32
+    app.dumpFiles(args["--output"].asString(), false, nsfilter);
+#else
+    app.dumpFiles(args["--dir"].asString(), redirect, nsfilter);
```
Why we use `--output` on WIN32 and `--dir` on linux/mac ?

MiguelRocha replied:
My type. Thanks

mgautierfr commented on src/zimdump.cpp:
```diff
@@ -90,7 +92,13 @@ class ZimDumper
+        verbose(false)
+      { }
+
+    ZimDumper(const std::string fname, bool titleSort)
```
You create a second creator but you never use it.
And I think we don't need it anyway.

MiguelRocha replied:
I removed it.

MiguelRocha commented:
> I think that you tried to keep the same structure and logic that the previous code but we should avoid that.
> 
> For example, you keep the "global" `pos` iterator but we shouldn't. Each method/command should do its own way. The `dump` method should use a iterator (cluster order), the `list` method should use a iterator (url order) and the list/dump with a selector should probably don't use a iterator but directly get an `Article`

Ok. I was trying to avoid making a big refactor in the code. But now, as you pointed, a lot of things start to not make sense (like the pos iterator).

MiguelRocha commented:
> @MiguelRocha The whole CI is failing, the discussions are not resumed. Might you please clean the situation?

The CI is now green. I'll apply the changes suggested by @mgautierfr. Meanwhile would be nice to have some input on the commands/API

mgautierfr commented:
> Meanwhile would be nice to have some input on the commands/API

As said in the review, please use the usage text defined in #70

mgautierfr commented on src/zimdump.cpp:
```diff
@@ -87,68 +87,72 @@ std::string urlEncode(const std::string& value, bool encodeReserved)
+    return m_file.getArticleByUrl(url);
+}
+
+zim::Article ZimDumper::getArticleByTitle(zim::size_type idx)
+{
+    return m_file.getArticleByTitle(idx);
+}
```
We should use `getArticle`. (url order)

MiguelRocha replied:
Done

mgautierfr commented on src/zimdump.cpp:
```diff
@@ -385,132 +367,143 @@ void ZimDumper::dumpFiles(const std::string& directory, bool symlinkdump)
+        std::unordered_map<std::string, std::function<void(ZimDumper&, decltype(args)&)>> dispatchtable = {
+            {"info",            subcmdInfo },
+            {"dump",            subcmdDump },
+            {"list",            subcmdList }
```
Is it missing the "ls" subcommand ? (or docopt detect it is a alias ?)

MiguelRocha replied:
Docopt doesnt seem to detect. I added it

mgautierfr commented on src/zimdump.cpp:
```diff
@@ -327,9 +306,12 @@ void ZimDumper::dumpFiles(const std::string& directory, bool symlinkdump)
+  for (zim::File::const_iterator it = m_file.begin(); it != m_file.end(); ++it)
   {
     std::string d = directory + SEPARATOR + it->getNamespace();
+    if (nsfilter(it->getNamespace()))
```
It is "funny", I would have written the other way (the filter must return `true` to keep the article). As for in python filter (https://docs.python.org/3/library/functions.html#filter) and c++ copy_if (https://en.cppreference.com/w/cpp/algorithm/copy)

But it works. No problem.

MiguelRocha replied:
Changed. I also think it makes a little bit more sense

MiguelRocha commented:
@kelson42 @mgautierfr I've implemented the changes from the code review and rebased the code. Let me know your comments.

MiguelRocha commented:
>     * I don't understand the difference between `list` and `ls`. Only one othem should here (`list`?).
> 
>     * I miss an introduction/description sentence(s) at the start of the usage(). Something like "Inspect and dump ZIM file content".
> 
>     * `zimdump 1.0` is not the proper version, should be the same version like other tools.
> 
>     * What does `[--]` means in `zimdump info [--] [--ns=N] <file>` (and in other usage() lines)?
> 
>     * `$ zimdump info --ns=42 vikidia_ca_all_maxi_2020-06.zim` does not return any error.
> 
>     * Because `zimdump dump` has two lines in the usage which work quite differently, I would propose to just move `zimdump dump (--idx=INDEX|(--url=URL [--ns=N])) [--] <file>` to `zimdump show (--idx=INDEX|(--url=URL [--ns=N])) [--] <file>`.
> 
> 
> @MiguelRocha Please resolve the conversation which should be resolved.

I removed "ls" alias as we agreed. All the other points are implemented. Let me know what you think

MiguelRocha commented:
> zimdump always return `0`. Please follow the usage text and return `!=0` if there is a error.
> 
> This pr is 21 commits long. Please squash the commits to have a commit per "logical change". If there is only one change that cannot be split, then use only one commit.

Added return errors.

I'll check the commit history and try to come up with a more readable commits but please keep in mind that since this is PR a major refactor to the code it is likely that we have a lot of commits.

MiguelRocha commented:
@mgautierfr I've squashed some commits. I removed commits that introduced just small changes/fixes to the code and I tried to keep the commits that introduce relevant (big) changes to the code base without losing a sense of "history".

mgautierfr commented on src/zimdump.cpp:
```diff
@@ -379,7 +375,7 @@ inline void write_to_file(const std::string &base, const std::string& path, cons
 
 
-void ZimDumper::dumpFiles(const std::string& directory, bool symlinkdump)
+int ZimDumper::dumpFiles(const std::string& directory, bool symlinkdump, std::function<bool (const char c)> nsfilter)
```
Do not make the method return a `int` if it always return 0.
It is useless and error prone (caller will assume you return something different than 0 if something goes wrong)

MiguelRocha replied:
I agree. I left `subcmdDumpAll` to return an int thou because it is part of the signature.

mgautierfr commented on src/zimdump.cpp:
```diff
@@ -463,134 +461,164 @@ void ZimDumper::dumpFiles(const std::string& directory, bool symlinkdump)
+    if (idx || url)
+    {
+        zim::Article article;
+        if (idx)
+        {
+            article = app.getArticle(args["--idx"].asLong());
+        } else if(url)
+        {
```
It is a matter of style, but I prefer the brace after the if :
```
if (idx) {
   ...
} else if (url) {
   ...
} else {
   ...
}
```

As it is done for the `for`, `try`, ...

MiguelRocha replied:
For me it is ok if everything follows the same convention. I tried to uniform for the other code also but I think we should have a separated PR to apply the code formatting rules (we can even have a clang-format file with our rules).

mgautierfr commented:
> I've squashed some commits. I removed commits that introduced just small changes/fixes to the code and I tried to keep the commits that introduce relevant (big) changes to the code base without losing a sense of "history".

That's good. I've check all commits individually but it what I wanted. Remove all iteration of the code review and keep different commits for different relevant changes.

MiguelRocha commented on src/zimdump.cpp:
```diff
@@ -463,134 +461,164 @@ void ZimDumper::dumpFiles(const std::string& directory, bool symlinkdump)
-    std::cout << "no checksum" << std::endl;
+    if (args["--ns"]) {
+        char ns = args["--ns"].asString().at(0);
+        if (app.hasNamespace(ns))
```
done

MiguelRocha commented:
> The code is functional. But I've few comments about the style.
> 
> Please try to keep the check consistent.
> 
> Subcommands are globally two parts :
> 
>     * Getting what to display/to do stuff on.
> 
>     * The action on the "object".
> 
> 
> Sometime, you are checking if the article is good in the first part before doing the action, and sometime you pass the article to the method and the method check it.

I decided to check on both since the methods can be used by other callers and it isn't a relevant performance decrease. 

I've also applied your suggestion in terms of style but since there is a lot to be said in terms of formatting I think we should have a separated PR to come up with the rules.

In terms of log history I didn't understand if the logs are OK or if you want more changes. I think every commit is separated in logical changes (that include changes that were requested in the code review process).

mgautierfr commented:
> I've also applied your suggestion in terms of style but since there is a lot to be said in terms of formatting I think we should have a separated PR to come up with the rules.

Yes, now we start to be several to hack on openzim/kiwix, I think it is time to have a coding style and a defined workflow about the PR and how to do.

I will do that and we will see then.

> In terms of log history I didn't understand if the logs are OK or if you want more changes. I think every commit is separated in logical changes (that include changes that were requested in the code review process).

The log history is good.

mgautierfr commented:
@MiguelRocha Please rebase on `origin/master`

MiguelRocha commented:
@mgautierfr  @kelson42  done. We are ready to merge.
